### PR TITLE
Rework journal abbreviation caching

### PR DESCRIPTION
**Pull request by tobiasdiez**
Rework the caching of the journal abbreviation list in order to reduce the memory footprint.
Instead of the plain in-memory cache, this PR now uses the [h2 MVStore](https://github.com/h2database/h2database) which is a light-weight file-based "database". I couldn't notice any performance degradation but a huge improvement in the memory department.

From:
image
To: 
image
Thus about 30% reduction. Which leads to a very reasonable memory usage: 400mb with a db of 1k entries. For larger databases the auto completion still leads to problems (see below, which was before the changes of this PR) - but this will be a new PR.
image
(The first two entries belong to the classloader and thus represent all libraries that we use). 

Further changes: 
- Removed the "IEEE abbreviation lists" and added them in the journal abbreviation repository. https://github.com/JabRef/abbrv.jabref.org/pull/64 https://github.com/JabRef/abbrv.jabref.org/pull/63   
- The current implementation uses a simple dictionary between full and abbreviated journal name for the built-in abbreviations. Thus, the "shortest abbreviation" or the "medline abbreviation" are ignored for the built-in list (there were like 2 journals that had this information...). If the user wants this, he can still add them by hand. 
- The abbreviation list is converted from the csv format to the MVStore format as part of the build process. This is able to use multiple csv files, which should aid #5749.

Further remarks:
- The performance now looks good enough to revisit https://github.com/JabRef/jabref/pull/5749 and try to include all 100k abbreviations in the built-in list.

- [ ] Change in CHANGELOG.md described (if applicable)
- [ ] Tests created for changes (if applicable)
- [ ] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, submitted a pull request to the documentation repository.

### CODE REVIEW

**Comment by jlaehne** [comment]
* There is an IEEE list in the abbreviation repository, so probably the "official IEEE" list from JabRef would have to be backported and merged with that one, making sure that changes listed in #5749 are also included. Then this list could be included in the automatic merge script, which would probably make more sense than providing a separate list in JabRef
* Instead JabRef should probably provide a list with dots and one without dots to choose from, currently they are created from two separate subsets of the repisitory and each has about 25k entries (the one without dots could be enriched by a merge with the other list after removing the dots).
* The 100k+ entries stem from including the webofscience lists from the repository. However, I would advise against including these lists in the official list, as detailed in https://github.com/JabRef/jabref/pull/5749#issuecomment-593148479 and https://github.com/JabRef/jabref/pull/5749#issuecomment-596382840
* The file merge should be automatically called from https://github.com/JabRef/jabref/pull/5749/files#diff-fde893913708a95fc89371a85acfcaf1 , which I guess is run during build and then the conversion to the MVStore should be called right afterwards (in the same script?)

**Comment by koppor** [comment]
> * I would propose to remove the "official IEEE abbreviation list" which changes e.g. "IEEE Transactions" to the bibtex string `#ieeetrans#`.

On the one hand, this feature makes JabRef special. On the other hand, this is a hidden gem.

This strongly relates to our improved typed `@STRING` handling: https://docs.jabref.org/fields/strings

In case users demand for this feature, they can "just" use the "IEEE abbreviation list", can't they?

> * The current implementation uses a simple dictionary between full and abbreviated journal name for the built-in abbreviations. Thus, the "shortest abbreviation" or the "medline abbreviation" are ignored (there were like 2 journals that had this information...).

Please do not drop this feature. I know that currently no one committed updated these abbreviations YET. Nevertheless, I believe in use cases where we really need that.

It should not be too hard to offer three different kind of journal name variants.

> The abbreviation list needs to be converted from the csv format to the MVStore format. This is done in JournalAbbreviationLoader#writeDefaultDatabase. I guess the best way would be to call this method as part of the build process or upon changing the abbreviation list. Any suggestions on how and when to do this?

+1 for build process. 

I plan to keep https://abbrv.jabref.org/ and enhance the abbreviation service offered there. The tooling to handle MVStore format is not as common as tooling for handling CSV files (looking at Python, Java, LibreOffice, GitHub browsing).

File: src/main/java/org/jabref/logic/journals/JournalAbbreviationLoader.java [hunk 1]
```diff
@@ -32,52 +35,75 @@
...
+                            .collect(Collectors.toMap(Abbreviation::getAbbreviation, Abbreviation::getName))
+            );
+
+            // Add all IEEE abbreviations
+            abbreviationToFull.putAll(
+                    getStandardIEEEAbbreviations()
+                            .stream()
+                            .collect(Collectors.toMap(Abbreviation::getAbbreviation, Abbreviation::getName))
+            );
+            //store.commit();
```
Comment (koppor) [inline src/main/java/org/jabref/logic/journals/JournalAbbreviationLoader.java:91]: Why is this disabled?

File: src/test/java/org/jabref/logic/integrity/IntegrityCheckTest.java [hunk 2]
```diff
@@ -368,7 +367,7 @@ void testEntryIsUnchangedAfterChecks() {
         new IntegrityCheck(context,
                 mock(FilePreferences.class),
                 createBibtexKeyPatternPreferences(),
-                           new JournalAbbreviationRepository(new Abbreviation("IEEE Software", "IEEE SW")), true, false)
+                new JournalAbbreviationRepository(), true, false)
```
Comment (koppor) [inline src/test/java/org/jabref/logic/integrity/IntegrityCheckTest.java:370]: Just stumbled over this - can we get rid of `true` and `false` somehow? - enums maybe?

**Comment by koppor** [comment]
Maybe, I should schedule a mob programming session at <https://github.com/remotemobprogramming/upcoming-sessions> to a) work on with-dot and dot-less and b) the three-state journal abbreviation?

File: src/main/java/org/jabref/logic/journals/JournalAbbreviationLoader.java (hunk 1 above)
Comment (tobiasdiez) [reply src/main/java/org/jabref/logic/journals/JournalAbbreviationLoader.java:91]: It's not need apparently (happens automatically if the store is closed).

File: src/test/java/org/jabref/logic/integrity/IntegrityCheckTest.java (hunk 2 above)
Comment (tobiasdiez) [reply src/test/java/org/jabref/logic/integrity/IntegrityCheckTest.java:370]: I don't see anything wrong with these booleans (and it's not related to this PR anyway)

**Comment by tobiasdiez** [comment]
Thanks @jlaehne and @koppor for the input. I've now updated the code and the PR description accordingly.

I ignored the doted version as well as the shortest abbreviation for now. In theory, this can be added but needs more work (need to worry about serialization in MvStore). Thus, this can be done at a later point if users really requests it (which I doubt to be honest).

File: buildSrc/build.gradle [hunk 3]
```diff
@@ -0,0 +1,20 @@
+apply plugin: 'java'
+
+repositories {
+    mavenLocal()
```
Comment (Siedlerchr) [inline buildSrc/build.gradle:4]: Maven local can be removed (only useful if you have gradle + maven builds in your system), will speed up the build.

File: src/main/java/org/jabref/gui/journals/AbbreviationViewModel.java [hunk 4]
```diff
@@ -80,16 +78,20 @@ public BooleanProperty isPseudoAbbreviationProperty() {
...
-        return Objects.hash(abbreviationObject);
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        AbbreviationViewModel that = (AbbreviationViewModel) o;
+        return getName().equals(that.getName()) &&
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/gui/journals/AbbreviationViewModel.java:89]: Use Objects.equals(...) to make it more consistent with other implementations

File: src/main/java/org/jabref/gui/journals/ManageJournalAbbreviationsView.java [hunk 5]
```diff
@@ -145,7 +147,11 @@ private void removeAbbreviation() {
 
     @FXML
     private void saveAbbreviationsAndCloseDialog() {
-        viewModel.saveEverythingAndUpdateAutoCompleter();
+        viewModel.save();
+
+        // Update journal abbreviation loader
+        Globals.journalAbbreviationRepository = JournalAbbreviationLoader.loadRepository(preferences.getJournalAbbreviationPreferences());
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/gui/journals/ManageJournalAbbreviationsView.java:153]: I do not really understand this line. Why is this necessary?

File: src/main/java/org/jabref/logic/journals/AbbreviationParser.java [hunk 6]
```diff
@@ -33,32 +30,20 @@
 
     public void readJournalListFromResource(String resourceFileName) {
         try {
-            URL url = Objects.requireNonNull(JournalAbbreviationRepository.class.getResource(Objects.requireNonNull(resourceFileName)));
-            readJournalList(new InputStreamReader(url.openStream(), StandardCharsets.UTF_8));
+            InputStream stream = JournalAbbreviationRepository.class.getResourceAsStream(resourceFileName);
+            readJournalList(new BufferedReader(new InputStreamReader(stream, StandardCharsets.UTF_8)));
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/logic/journals/AbbreviationParser.java:34]: Possible resource leak? I don't see where the reader is closed.

File: src/main/java/org/jabref/logic/journals/JournalAbbreviationLoader.java [hunk 7]
```diff
@@ -14,58 +16,42 @@
...
-        if (journalAbbreviationPreferences.useIEEEAbbreviations()) {
-            journalAbbrev.addEntries(getOfficialIEEEAbbreviations());
-        } else {
-            journalAbbrev.addEntries(getStandardIEEEAbbreviations());
+    public static JournalAbbreviationRepository loadRepository(JournalAbbreviationPreferences journalAbbreviationPreferences) {
+        JournalAbbreviationRepository repository;
+        // Initialize with built-in list
+        try {
+            Path tempJournalList = Files.createTempDirectory("journal").resolve("journalList.mv");
+            Files.copy(JournalAbbreviationRepository.class.getResourceAsStream("/journals/journalList.mv"), tempJournalList);
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/logic/journals/JournalAbbreviationLoader.java:50]: We have to check if this also works in the jlink/jpackage version.

File: src/main/java/org/jabref/logic/journals/JournalAbbreviationRepository.java [hunk 8]
```diff
@@ -36,65 +42,89 @@ private static boolean isMatchedAbbreviated(String name, Abbreviation abbreviati
...
 
     /**
      * Returns true if the given journal name is in its abbreviated form (e.g. Phys. Rev. Lett.). The test is strict,
      * i.e. journals whose abbreviation is the same as the full name are not considered
      */
     public boolean isAbbreviatedName(String journalName) {
-        return abbreviations.stream().anyMatch(abbreviation -> isMatchedAbbreviated(journalName.trim(), abbreviation));
+        String journal = journalName.trim();
+
+        boolean isAbbreviated = customAbbreviations.stream().anyMatch(abbreviation -> isMatchedAbbreviated(journal, abbreviation));
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/logic/journals/JournalAbbreviationRepository.java:67]: Combine both returns with an OR?

File: src/main/java/org/jabref/gui/journals/ManageJournalAbbreviationsView.java (hunk 5 above)
Comment (tobiasdiez) [reply src/main/java/org/jabref/gui/journals/ManageJournalAbbreviationsView.java:153]: If the user changes the abbreviations, these changes must be reflected in the global repository.

**Comment by tobiasdiez** [comment]
Thanks for the review @Siedlerchr. Feedback is now included.

**Comment by Siedlerchr** [comment]
Codewise looks good, we just have to test the changes now in the jlink/jpackage variant to make sure the classpath resources etc are correctly found

**Comment by koppor** [comment]
@Siedlerchr That means, the only thing left to test is to download the binary from https://builds.jabref.org/pull/6304/merge/ and to execute it?

**Comment by koppor** [comment]
Does not work with `jabref-authors.bib`.

1. Open jabref-authors.bib
2. Abbreviate Journal names --> "MEDLINE"
3. 9 changes

   One wrong: grafik

4. Save
5. changes in bib file are there
6. "Unnabbreviate journal names"
7. 11 changes (according to pop up) (or 3 changes)
8. Save
9. No changes in the bib file are shown

Is "MEDLINE" now forbidden?

**Comment by koppor** [comment]
In short: When "unabbreviating", no changes are written. Just load `jabref-authors.bib` and execute "Unnabreviate" and save. No changes in .bib file.

**Comment by tobiasdiez** [comment]
Good observation, there was indeed a small bug in the code. Unnabbreviation of names with dots should work now. As I said above, the built-in list does no longer contain names without dots, so medline unabbreviation does indeed no longer work (something for a new PR in my opinion).

**Comment by tobiasdiez** [comment]
Thanks @calixtus for the feedback. Your suggestions are now included.

I'll merge this now. @koppor feel free to add additional comments if you think there is something that needs further work.

**Comment by Siedlerchr** [comment]
The files are generated every time on build time? So we probably need to add the -mv file to the gitignore file. Just noticed that

**Comment by tobiasdiez** [comment]
At the moment yes, but I would propose to only add the mv file and remove the cls one: https://github.com/JabRef/jabref/pull/5749#issuecomment-622366394
